File: backend/layers/shared/tourney.py
"""
Tournament state management and business logic
"""
from typing import Dict, List, Optional, Tuple
from decimal import Decimal


class TourneyStatus:
    """Tournament status constants"""
    WAITING = "waiting"
    STARTING = "starting"
    IN_PROGRESS = "in_progress"
    BETWEEN_GAMES = "between_games"
    COMPLETED = "completed"


class Seat:
    """Represents a tournament seat"""
    def __init__(self, position: int, player_id: str = None, player_name: str = None,
                 connection_id: str = None, score: int = 0, games_won: int = 0,
                 last_game_points: int = 0, ready: bool = False):
        self.position = position
        self.player_id = player_id
        self.player_name = player_name
        self.connection_id = connection_id
        self.score = score
        self.games_won = games_won
        self.last_game_points = last_game_points
        self.ready = ready

    def to_dict(self) -> Dict:
        """Convert to DynamoDB-compatible dict"""
        return {
            'position': self.position,
            'playerId': self.player_id,
            'playerName': self.player_name,
            'connectionId': self.connection_id,
            'score': self.score,
            'gamesWon': self.games_won,
            'lastGamePoints': self.last_game_points,
            'ready': self.ready
        }

    @classmethod
    def from_dict(cls, data: Dict) -> 'Seat':
        """Create from DynamoDB dict"""
        return cls(
            position=data['position'],
            player_id=data.get('playerId'),
            player_name=data.get('playerName'),
            connection_id=data.get('connectionId'),
            score=int(data.get('score', 0)),
            games_won=int(data.get('gamesWon', 0)),
            last_game_points=int(data.get('lastGamePoints', 0)),
            ready=data.get('ready', False)
        )

    def is_occupied(self) -> bool:
        """Check if seat is occupied"""
        return self.player_id is not None

    def is_empty(self) -> bool:
        """Check if seat is empty"""
        return self.player_id is None


class Tourney:
    """Tournament state and business logic"""
    GLOBAL_ID = "global"
    TARGET_SCORE = 21
    SEATS_COUNT = 4

    def __init__(self, tourney_id: str = GLOBAL_ID):
        self.tourney_id = tourney_id
        self.status = TourneyStatus.WAITING
        self.target_score = self.TARGET_SCORE
        self.seats: List[Seat] = [Seat(i) for i in range(self.SEATS_COUNT)]
        self.current_game = None
        self.game_history = []

    @classmethod
    def from_dynamo(cls, item: Dict) -> 'Tourney':
        """Create tournament from DynamoDB item"""
        tourney = cls(item['tourneyId'])
        tourney.status = item.get('status', TourneyStatus.WAITING)
        tourney.target_score = int(item.get('targetScore', cls.TARGET_SCORE))

        # Parse seats
        seats_data = item.get('seats', [])
        tourney.seats = [Seat.from_dict(s) for s in seats_data]

        # Ensure we always have 4 seats
        while len(tourney.seats) < cls.SEATS_COUNT:
            tourney.seats.append(Seat(len(tourney.seats)))

        tourney.current_game = item.get('currentGame')
        tourney.game_history = item.get('gameHistory', [])

        return tourney

    def to_dynamo(self) -> Dict:
        """Convert to DynamoDB item"""
        return {
            'tourneyId': self.tourney_id,
            'status': self.status,
            'targetScore': self.target_score,
            'seats': [s.to_dict() for s in self.seats],
            'currentGame': self.current_game,
            'gameHistory': self.game_history
        }

    def claim_seat(self, player_id: str, player_name: str, connection_id: str,
                   seat_position: Optional[int] = None) -> Tuple[bool, str, Optional[int]]:
        """
        Claim a seat in the tournament

        Returns: (success: bool, error_message: str, claimed_position: int)
        """
        # Check if tourney is in valid state for claiming
        if self.status not in [TourneyStatus.WAITING, TourneyStatus.STARTING]:
            return False, "TOURNEY_IN_PROGRESS", None

        # Check if player already has a seat
        existing_seat = self.get_seat_by_player(player_id)
        if existing_seat is not None:
            # Update connection ID for reconnection
            existing_seat.connection_id = connection_id
            return True, "", existing_seat.position

        # Find available seat
        if seat_position is not None:
            # Specific seat requested
            if seat_position < 0 or seat_position >= self.SEATS_COUNT:
                return False, "INVALID_SEAT", None

            if self.seats[seat_position].is_occupied():
                return False, "SEAT_TAKEN", None

            target_seat = self.seats[seat_position]
        else:
            # First available seat
            target_seat = self.get_first_empty_seat()
            if target_seat is None:
                return False, "TOURNEY_FULL", None

        # Claim the seat
        target_seat.player_id = player_id
        target_seat.player_name = player_name
        target_seat.connection_id = connection_id
        target_seat.score = 0
        target_seat.games_won = 0
        target_seat.last_game_points = 0
        target_seat.ready = False

        # Update status if all seats filled
        if self.get_occupied_count() == self.SEATS_COUNT:
            if self.status == TourneyStatus.WAITING:
                self.status = TourneyStatus.STARTING

        return True, "", target_seat.position

    def leave_tourney(self, player_id: str) -> Tuple[bool, str]:
        """
        Leave the tournament

        Returns: (success: bool, error_message: str)
        """
        seat = self.get_seat_by_player(player_id)
        if seat is None:
            return False, "NOT_IN_TOURNEY"

        # If tournament hasn't started, free the seat completely
        if self.status in [TourneyStatus.WAITING, TourneyStatus.STARTING]:
            seat.player_id = None
            seat.player_name = None
            seat.connection_id = None
            seat.score = 0
            seat.games_won = 0
            seat.last_game_points = 0
            seat.ready = False

            # Update status if no longer full
            if self.get_occupied_count() < self.SEATS_COUNT:
                self.status = TourneyStatus.WAITING

            return True, ""

        # If tournament in progress, mark as disconnected (TODO: bot replacement)
        # For now, just return error
        return False, "TOURNEY_IN_PROGRESS"

    def set_ready(self, player_id: str, ready: bool = True) -> Tuple[bool, str]:
        """
        Mark player as ready

        Returns: (success: bool, error_message: str)
        """
        seat = self.get_seat_by_player(player_id)
        if seat is None:
            return False, "NOT_IN_TOURNEY"

        # Can only ready up in starting or between_games states
        if self.status not in [TourneyStatus.STARTING, TourneyStatus.BETWEEN_GAMES]:
            return False, "INVALID_STATE"

        seat.ready = ready

        # Check if all players ready
        if self.are_all_ready():
            # Start game or next game
            if self.status == TourneyStatus.STARTING:
                self.status = TourneyStatus.IN_PROGRESS
                # TODO: Initialize first game
            elif self.status == TourneyStatus.BETWEEN_GAMES:
                self.status = TourneyStatus.IN_PROGRESS
                # TODO: Initialize next game

        return True, ""

    def get_seat_by_player(self, player_id: str) -> Optional[Seat]:
        """Get seat occupied by player"""
        for seat in self.seats:
            if seat.player_id == player_id:
                return seat
        return None

    def get_first_empty_seat(self) -> Optional[Seat]:
        """Get first available empty seat"""
        for seat in self.seats:
            if seat.is_empty():
                return seat
        return None

    def get_occupied_count(self) -> int:
        """Count occupied seats"""
        return sum(1 for s in self.seats if s.is_occupied())

    def get_ready_count(self) -> int:
        """Count ready players"""
        return sum(1 for s in self.seats if s.is_occupied() and s.ready)

    def are_all_ready(self) -> bool:
        """Check if all occupied seats are ready"""
        occupied = [s for s in self.seats if s.is_occupied()]
        if not occupied:
            return False
        return all(s.ready for s in occupied)

    def to_client_state(self) -> Dict:
        """Convert to client-friendly state"""
        return {
            'status': self.status,
            'seats': [
                {
                    'position': s.position,
                    'playerName': s.player_name,
                    'score': s.score,
                    'gamesWon': s.games_won,
                    'ready': s.ready
                }
                for s in self.seats
            ],
            'targetScore': self.target_score,
            'currentGameNumber': len(self.game_history) + (1 if self.current_game else 0),
            'readyCount': self.get_ready_count()
        }

File: packages/game-logic/src/tourney.ts
export enum TourneyStatus {
  WAITING = "waiting",
  STARTING = "starting",
  IN_PROGRESS = "in_progress",
  BETWEEN_GAMES = "between_games",
  COMPLETED = "completed",
}

export interface SeatData {
  position: number;
  playerId: string | null;
  playerName: string | null;
  connectionId: string | null;
  score: number;
  gamesWon: number;
  lastGamePoints: number;
  ready: boolean;
  disconnectedAt?: number;
  isBot: boolean;
  botProfile?: string;
}

export class Seat {
  position: number;
  playerId: string | null;
  playerName: string | null;
  connectionId: string | null;
  score: number;
  gamesWon: number;
  lastGamePoints: number;
  ready: boolean;
  disconnectedAt?: number;
  isBot: boolean;
  botProfile?: string;

  constructor(position: number, data?: Partial<SeatData>) {
    this.position = position;
    this.playerId = data?.playerId ?? null;
    this.playerName = data?.playerName ?? null;
    this.connectionId = data?.connectionId ?? null;
    this.score = data?.score ?? 0;
    this.gamesWon = data?.gamesWon ?? 0;
    this.lastGamePoints = data?.lastGamePoints ?? 0;
    this.ready = data?.ready ?? false;
    this.disconnectedAt = data?.disconnectedAt;
    this.isBot = data?.isBot ?? false;
    this.botProfile = data?.botProfile;
  }

  isOccupied(): boolean {
    return this.playerId !== null;
  }

  isEmpty(): boolean {
    return this.playerId === null;
  }

  clear(): void {
    this.playerId = null;
    this.playerName = null;
    this.connectionId = null;
    this.score = 0;
    this.gamesWon = 0;
    this.lastGamePoints = 0;
    this.ready = false;
    this.isBot = false;
    this.botProfile = undefined;
    this.disconnectedAt = undefined;
  }

  toDict(): SeatData {
    const result: SeatData = {
      position: this.position,
      playerId: this.playerId,
      playerName: this.playerName,
      connectionId: this.connectionId,
      score: this.score,
      gamesWon: this.gamesWon,
      lastGamePoints: this.lastGamePoints,
      ready: this.ready,
      isBot: this.isBot,
    };
    if (this.disconnectedAt !== undefined)
      result.disconnectedAt = this.disconnectedAt;
    if (this.botProfile !== undefined) result.botProfile = this.botProfile;
    return result;
  }

  static fromDict(data: SeatData): Seat {
    return new Seat(data.position, data);
  }
}

type Result = [success: boolean, error: string];
type ClaimResult = [success: boolean, error: string, position: number | null];

export class Tourney {
  static readonly GLOBAL_ID = "global";
  static readonly TARGET_SCORE = 21;
  static readonly SEATS_COUNT = 4;

  tourneyId: string;
  status: TourneyStatus;
  targetScore: number;
  seats: Seat[];
  currentGame: Record<string, unknown> | null;
  gameHistory: Record<string, unknown>[];

  constructor(tourneyId = Tourney.GLOBAL_ID) {
    this.tourneyId = tourneyId;
    this.status = TourneyStatus.WAITING;
    this.targetScore = Tourney.TARGET_SCORE;
    this.seats = Array.from({ length: Tourney.SEATS_COUNT }, (_, i) => new Seat(i));
    this.currentGame = null;
    this.gameHistory = [];
  }

  static fromDynamo(item: Record<string, unknown>): Tourney {
    const tourney = new Tourney(item.tourneyId as string);
    tourney.status =
      (item.status as TourneyStatus) ?? TourneyStatus.WAITING;
    tourney.targetScore = Number(item.targetScore ?? Tourney.TARGET_SCORE);

    const seatsData = (item.seats as SeatData[]) ?? [];
    tourney.seats = seatsData.map((s) => Seat.fromDict(s));
    while (tourney.seats.length < Tourney.SEATS_COUNT) {
      tourney.seats.push(new Seat(tourney.seats.length));
    }

    tourney.currentGame =
      (item.currentGame as Record<string, unknown>) ?? null;
    tourney.gameHistory =
      (item.gameHistory as Record<string, unknown>[]) ?? [];

    return tourney;
  }

  toDynamo(): Record<string, unknown> {
    return {
      tourneyId: this.tourneyId,
      status: this.status,
      targetScore: this.targetScore,
      seats: this.seats.map((s) => s.toDict()),
      currentGame: this.currentGame,
      gameHistory: this.gameHistory,
    };
  }

  claimSeat(
    playerId: string,
    playerName: string,
    connectionId: string,
    seatPosition?: number,
  ): ClaimResult {
    if (
      this.status !== TourneyStatus.WAITING &&
      this.status !== TourneyStatus.STARTING
    ) {
      return [false, "TOURNEY_IN_PROGRESS", null];
    }

    // Check if player already has a seat
    const existing = this.getSeatByPlayer(playerId);
    if (existing) {
      existing.connectionId = connectionId;
      return [true, "", existing.position];
    }

    let targetSeat: Seat | null;
    if (seatPosition !== undefined) {
      if (seatPosition < 0 || seatPosition >= Tourney.SEATS_COUNT) {
        return [false, "INVALID_SEAT", null];
      }
      if (this.seats[seatPosition].isOccupied()) {
        return [false, "SEAT_TAKEN", null];
      }
      targetSeat = this.seats[seatPosition];
    } else {
      targetSeat = this.getFirstEmptySeat();
      if (!targetSeat) return [false, "TOURNEY_FULL", null];
    }

    targetSeat.playerId = playerId;
    targetSeat.playerName = playerName;
    targetSeat.connectionId = connectionId;
    targetSeat.score = 0;
    targetSeat.gamesWon = 0;
    targetSeat.lastGamePoints = 0;
    targetSeat.ready = false;

    if (this.getOccupiedCount() === Tourney.SEATS_COUNT) {
      if (this.status === TourneyStatus.WAITING) {
        this.status = TourneyStatus.STARTING;
      }
    }

    return [true, "", targetSeat.position];
  }

  leaveTourney(playerId: string): Result {
    const seat = this.getSeatByPlayer(playerId);
    if (!seat) return [false, "NOT_IN_TOURNEY"];

    if (
      this.status === TourneyStatus.WAITING ||
      this.status === TourneyStatus.STARTING
    ) {
      seat.clear();
      if (this.getOccupiedCount() < Tourney.SEATS_COUNT) {
        this.status = TourneyStatus.WAITING;
      }
      return [true, ""];
    }

    return [false, "TOURNEY_IN_PROGRESS"];
  }

  addBot(seatPosition: number, botProfile?: string): Result {
    if (
      this.status !== TourneyStatus.WAITING &&
      this.status !== TourneyStatus.STARTING
    ) {
      return [false, "TOURNEY_IN_PROGRESS"];
    }
    if (seatPosition < 0 || seatPosition >= Tourney.SEATS_COUNT) {
      return [false, "INVALID_SEAT"];
    }
    const seat = this.seats[seatPosition];
    if (seat.isOccupied()) return [false, "SEAT_TAKEN"];

    const botId = `bot_${crypto.randomUUID().slice(0, 8)}`;
    seat.playerId = botId;
    seat.playerName = `Bot_${seatPosition + 1}`;
    seat.connectionId = null;
    seat.score = 0;
    seat.gamesWon = 0;
    seat.lastGamePoints = 0;
    seat.ready = true;
    seat.isBot = true;
    seat.botProfile = botProfile;

    if (this.getOccupiedCount() === Tourney.SEATS_COUNT) {
      if (this.status === TourneyStatus.WAITING) {
        this.status = TourneyStatus.STARTING;
      }
    }

    return [true, ""];
  }

  kickBot(seatPosition: number): Result {
    if (
      this.status !== TourneyStatus.WAITING &&
      this.status !== TourneyStatus.STARTING
    ) {
      return [false, "TOURNEY_IN_PROGRESS"];
    }
    if (seatPosition < 0 || seatPosition >= Tourney.SEATS_COUNT) {
      return [false, "INVALID_SEAT"];
    }
    const seat = this.seats[seatPosition];
    if (!seat.isOccupied()) return [false, "SEAT_EMPTY"];
    if (!seat.isBot) return [false, "NOT_A_BOT"];

    seat.clear();
    if (this.getOccupiedCount() < Tourney.SEATS_COUNT) {
      this.status = TourneyStatus.WAITING;
    }

    return [true, ""];
  }

  setReady(playerId: string, ready = true): Result {
    const seat = this.getSeatByPlayer(playerId);
    if (!seat) return [false, "NOT_IN_TOURNEY"];

    if (
      this.status !== TourneyStatus.STARTING &&
      this.status !== TourneyStatus.BETWEEN_GAMES
    ) {
      return [false, "INVALID_STATE"];
    }

    seat.ready = ready;

    if (this.areAllReady()) {
      this.status = TourneyStatus.IN_PROGRESS;
    }

    return [true, ""];
  }

  completeGame(winOrder: number[]): [success: boolean, tournamentComplete: boolean] {
    if (!this.currentGame) return [false, false];

    const pointsAwarded = [4, 2, 1, 0];
    for (let i = 0; i < winOrder.length; i++) {
      const seat = this.seats[winOrder[i]];
      const points = pointsAwarded[i];
      seat.score += points;
      seat.lastGamePoints = points;
      if (i === 0) seat.gamesWon++;
    }

    this.gameHistory.push({
      gameNumber: this.gameHistory.length + 1,
      winOrder,
      pointsAwarded,
    });

    this.currentGame = null;

    const maxScore = Math.max(...this.seats.map((s) => s.score));
    const tournamentComplete = maxScore >= this.targetScore;

    this.status = tournamentComplete
      ? TourneyStatus.COMPLETED
      : TourneyStatus.BETWEEN_GAMES;

    return [true, tournamentComplete];
  }

  getLeaderboard(): Array<{
    position: number;
    playerName: string | null;
    totalScore: number;
    lastGamePoints: number;
    gamesWon: number;
  }> {
    return this.seats
      .filter((s) => s.isOccupied())
      .map((s) => ({
        position: s.position,
        playerName: s.playerName,
        totalScore: s.score,
        lastGamePoints: s.lastGamePoints,
        gamesWon: s.gamesWon,
      }))
      .sort((a, b) => b.totalScore - a.totalScore);
  }

  toClientState(): Record<string, unknown> {
    return {
      status: this.status,
      seats: this.seats.map((s) => ({
        position: s.position,
        playerName: s.playerName,
        score: s.score,
        gamesWon: s.gamesWon,
        ready: s.ready,
        isBot: s.isBot,
      })),
      targetScore: this.targetScore,
      currentGameNumber:
        this.gameHistory.length + (this.currentGame ? 1 : 0),
      readyCount: this.getReadyCount(),
    };
  }

  // ── Helpers ──

  getSeatByPlayer(playerId: string): Seat | null {
    return this.seats.find((s) => s.playerId === playerId) ?? null;
  }

  getFirstEmptySeat(): Seat | null {
    return this.seats.find((s) => s.isEmpty()) ?? null;
  }

  getOccupiedCount(): number {
    return this.seats.filter((s) => s.isOccupied()).length;
  }

  getReadyCount(): number {
    return this.seats.filter((s) => s.isOccupied() && s.ready).length;
  }

  areAllReady(): boolean {
    const occupied = this.seats.filter((s) => s.isOccupied());
    if (occupied.length === 0) return false;
    return occupied.every((s) => s.ready);
  }
}
